### PR TITLE
Cleanup detection and avoid calling endpoints that are not present

### PR DESCRIPTION
**Pull request by bdraco**
- support passing in shared httpx client
- get pylint rating to 10/10
```
pylint envoy_reader/envoy_reader.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### CODE REVIEW

**Comment by bdraco** [comment]
@gtdiehl I remember from https://github.com/home-assistant/core/pull/42857 that you have write access to this repo so hopefully you can take a look.

Thanks!

**Comment by gtdiehl** [comment]
@bdraco I know I've been silent for a bit. COVID, family issues, death, and other stuff have side-lined me. I will take a look this week I promise! I'll see about testing this against the various platforms. I should be able to get access to Envoys running <3.9, >=3.9, and >=4.x.

**Comment by bdraco** [comment]
> @bdraco I know I've been silent for a bit. COVID, family issues, death, and other stuff have side-lined me. I will take a look this tomorrow I promise! I'll see about testing this against the various platforms. I should be able to get access to Envoys running <3.9, >=3.9, and >=4.x.

Sorry man that sucks that the world hasn't been kind lately. Hope it gets better for you soon.

Thanks for taking a look 👍

**Comment by gtdiehl** [comment]
@bdraco Thank you!

I was able to test this on my Envoy running 3.15 and another running 4.x. I still want to run it against <3.9 as these don't have the same pages. I wasn't able to find one today. Hopefully I can find one tomorrow.

File: envoy_reader/envoy_reader.py [hunk 1]
```diff
@@ -133,8 +133,6 @@ async def getData(self):  # pylint: disable=invalid-name
                         response.raise_for_status()
                 except (httpcore.RemoteProtocolError, httpx.RemoteProtocolError):
                     continue
-                except httpx.HTTPError:
-                    response.raise_for_status()
```
Comment (bdraco) [inline envoy_reader/envoy_reader.py]: Fixes
```
2021-04-08 02:04:54 ERROR (MainThread) [homeassistant.components.enphase_envoy] Unexpected error fetching envoy {name} data: local variable 'response' referenced before assignment
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/httpx/_exceptions.py", line 326, in map_exceptions
    yield
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1492, in _send_single_request
    (status_code, headers, stream, ext) = await transport.arequest(
  File "/usr/local/lib/python3.8/site-packages/httpx/_transports/default.py", line 169, in arequest
    return await self._pool.arequest(
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/connection_pool.py", line 218, in arequest
    response = await connection.arequest(
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/connection.py", line 106, in arequest
    return await self.connection.arequest(method, url, headers, stream, ext)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/http11.py", line 72, in arequest
    ) = await self._receive_response(timeout)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/http11.py", line 133, in _receive_response
    event = await self._receive_event(timeout)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/http11.py", line 172, in _receive_event
    data = await self.socket.read(self.READ_NUM_BYTES, timeout)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_backends/asyncio.py", line 157, in read
    raise
  File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_exceptions.py", line 12, in map_exceptions
    raise to_exc(exc) from None
httpcore.ReadError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/envoy_reader/envoy_reader.py", line 125, in getData
    response = await client.get(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1539, in get
    return await self.request(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1361, in request
    response = await self.send(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1396, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1434, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1466, in _send_handling_redirects
    response = await self._send_single_request(request, timeout)
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1492, in _send_single_request
    (status_code, headers, stream, ext) = await transport.arequest(
  File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.8/site-packages/httpx/_exceptions.py", line 343, in map_exceptions
    raise mapped_exc(message, **kwargs) from exc  # type: ignore
httpx.ReadError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 175, in _async_refresh
    self.data = await self._async_update_data()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 145, in _async_update_data
    return await self.update_method()
  File "/usr/src/homeassistant/homeassistant/components/enphase_envoy/__init__.py", line 52, in async_update_data
    await envoy_reader.getData()
  File "/usr/local/lib/python3.8/site-packages/envoy_reader/envoy_reader.py", line 137, in getData
    response.raise_for_status()
UnboundLocalError: local variable 'response' referenced before assignment
```

Comment (gtdiehl) [reply envoy_reader/envoy_reader.py]: @bdraco I was testing this against an Envoy running `D5.0.49` and noticed the local variable error, but also noticed that I was not able to retrieve the Inverter data. This envoy has the standard login of envoy for the username and the last 6 digits of the serial number for the password. I tried entering the credentials manually but I got the same response.

I didn't run it in Home Assistant just did a quick test from the command line before moving to the integration with HA.

```
Reading...
production:              local variable 'production' referenced before assignment
consumption:             511
daily_production:        local variable 'daily_production' referenced before assignment
daily_consumption:       1982
seven_days_production:   local variable 'seven_days_production' referenced before assignment
seven_days_consumption:  167894
lifetime_production:     local variable 'lifetime_production' referenced before assignment
lifetime_consumption:    42204657
inverters_production:    Inverter data not available for your Envoy device.
```

Comment (bdraco) [reply envoy_reader/envoy_reader.py]: I moved it before the class. It should work correctly now

**Comment by gtdiehl** [comment]
@bdraco I tested on several Enphase Envoys, R3.7.31, R3.17.3, D4.2.27, and D5.0.49. Everything looks good! I want to run a couple more tests but need the sun shining. I just want to test when the Envoy is and is not using a Production CT.

If everything looks good tomorrow and I merge the code into the main branch, do you need a new release on PyPi as well?

**Comment by bdraco** [comment]
> @bdraco I tested on several Enphase Envoys, R3.7.31, R3.17.3, D4.2.27, and D5.0.49. Everything looks good! I want to run a couple more tests but need the sun shining. I just want to test when the Envoy is and is not using a Production CT.
> 
> If everything looks good tomorrow and I merge the code into the master branch, do you need a new release on PyPi as well?

Excellent. Thanks for testing.

A new release would be great as I would like to get https://github.com/home-assistant/core/pull/48709 moved forward

**Comment by gtdiehl** [comment]
@bdraco I'm sorry but I found a problem. I found an Evoy running D4.x firmware that does not have a Production CT (eim activeCount = 0), and I'm not getting back Daily, Weekly, or Lifetime production values. 

I'll try  and debug it a bit more later tonight.

Here is the output from the envoy_reader.py and JSON
```
Reading...
production:              5891
consumption:             5806
daily_production:        'NoneType' object has no attribute 'json'
daily_consumption:       0
seven_days_production:   'NoneType' object has no attribute 'json'
seven_days_consumption:  0
lifetime_production:     'NoneType' object has no attribute 'json'
lifetime_consumption:    0
inverters_production:    {...}
```

/production.json
```
{"production":[{"type":"inverters","activeCount":34,"readingTime":1618005986,"wNow":5891,"whLifetime":10279087},{"type":"eim","activeCount":0,"measurementType":"production","readingTime":1618006624,"wNow":5814.869,"whLifetime":0.0,"varhLeadLifetime":0.0,"varhLagLifetime":0.0,"vahLifetime":0.0,"rmsCurrent":47.811,"rmsVoltage":243.277,"reactPwr":465.213,"apprntPwr":5816.157,"pwrFactor":1.0,"whToday":0.0,"whLastSevenDays":0.0,"vahToday":0.0,"varhLeadToday":0.0,"varhLagToday":0.0}],"consumption":[{"type":"eim","activeCount":0,"measurementType":"total-consumption","readingTime":1618006624,"wNow":5811.099,"whLifetime":0.0,"varhLeadLifetime":0.0,"varhLagLifetime":0.0,"vahLifetime":0.0,"rmsCurrent":47.534,"rmsVoltage":243.146,"reactPwr":-465.213,"apprntPwr":11557.595,"pwrFactor":0.5,"whToday":0.0,"whLastSevenDays":0.0,"vahToday":0.0,"varhLeadToday":0.0,"varhLagToday":0.0},{"type":"eim","activeCount":0,"measurementType":"net-consumption","readingTime":1618006624,"wNow":-3.769,"whLifetime":0.0,"varhLeadLifetime":0.0,"varhLagLifetime":0.0,"vahLifetime":0.0,"rmsCurrent":0.278,"rmsVoltage":243.015,"reactPwr":-0.0,"apprntPwr":33.697,"pwrFactor":0.0,"whToday":0,"whLastSevenDays":0,"vahToday":0,"varhLeadToday":0,"varhLagToday":0}],"storage":[{"type":"acb","activeCount":0,"readingTime":0,"wNow":0,"whNow":0,"state":"idle"}]}
```

/api/v1/production
```
{
  "wattHoursToday": 17920,
  "wattHoursSevenDays": 276614,
  "wattHoursLifetime": 10279087,
  "wattsNow": 5894
}
```

**Comment by bdraco** [comment]
Does that one work with the current master or is that a regression?

**Comment by gtdiehl** [comment]
It's a regression with this PR.

This is the ouput using the same Envoy with the current master code
```
Reading...
production:              5680
consumption:             5292
daily_production:        29477
daily_consumption:       0
seven_days_production:   279353
seven_days_consumption:  0
lifetime_production:     10290644
lifetime_consumption:    0
inverters_production:    {...}
```

**Comment by bdraco** [comment]
Thanks. I'll start some tests tomorrow with the data dump you provided

**Comment by gtdiehl** [comment]
From what I can tell, when the Envoy type is `PC` the `endpoint_production_v1_results` never gets populated, during `getData()`.

We would still need to retrieve the `endpoint_production_v1_results` when Production Metering is disabled as the JSON data for Daily, Weekly, and Lifetime Production is not available from `/production.json` but from `/api/v1/production`

Hope this helps! 😃

**Comment by bdraco** [comment]
Working on adding test fixtures for `3.17.3	3.9.36	4.2.27	5.0.49` 

I'll push something later today

**Comment by bdraco** [comment]
@gtdiehl Should be fixed now

**Comment by gtdiehl** [comment]
I've tested against different firmware and configurations. Looks good to me!

**Comment by bdraco** [comment]
Thanks!

**Pull request merged**